File: apps/yryvu-app/src/ipc/pr_detail.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

import type {
  CiStatus,
  Label,
  PullRequestState,
  ReviewDecision,
  UserInfo,
} from "./integrationStorage";

/// GitLab project setting controlling how merges are committed.
/// Mirrors the backend `ProjectMergeMethod` enum (camelCase wire form).
export type ProjectMergeMethod = "merge" | "rebaseMerge" | "ff";

/// GitLab project setting controlling whether commits are squashed at
/// merge time. Mirrors the backend `ProjectSquashOption` enum.
export type ProjectSquashOption =
  | "never"
  | "always"
  | "defaultOff"
  | "defaultOn";

/// GitLab-only project merge config powering the merge form's
/// radio gating + squash checkbox availability. `null` for GitHub /
/// Gitea — the form falls back to unrestricted controls.
export interface ProjectMergeSettings {
  mergeMethod: ProjectMergeMethod;
  squashOption: ProjectSquashOption;
  removeSourceBranchAfterMergeDefault: boolean;
  allowMergeOnSkippedPipeline: boolean;
}

/**
 * Extended PR record — superset of `PullRequestSummary` with body
 * (markdown), mergeability, and aggregate counts. Powers the 4-tab
 * detail panel (#91).
 */
export interface PullRequestDetail {
  number: number;
  title: string;
  state: PullRequestState;
  draft: boolean;
  author: UserInfo;
  createdAt: string;
  updatedAt: string;
  closedAt: string | null;
  mergedAt: string | null;
  htmlUrl: string;
  baseRef: string;
  headRef: string;
  headSha: string;
  body: string;
  labels: Label[];
  assignees: UserInfo[];
  requestedReviewers: UserInfo[];
  milestone: string | null;
  /// `true` = mergeable, `false` = conflicting, `null` = still computing.
  mergeable: boolean | null;
  /// Verbose verb: `clean | dirty | blocked | unstable | behind |
  /// has_hooks | unknown`. Drives Merge-button enablement.
  mergeableState: string | null;
  additions: number;
  deletions: number;
  changedFiles: number;
  comments: number;
  reviewDecision: ReviewDecision | null;
  ciStatus: CiStatus | null;
  /// GitLab-only project merge config; `null` for GitHub / Gitea.
  projectSettings: ProjectMergeSettings | null;
}

export interface PrCommit {
  sha: string;
  shortSha: string;
  message: string;
  author: UserInfo;
  date: string;
}

export interface PrFile {
  filename: string;
  /// `added | modified | removed | renamed | copied | changed | unchanged`.
  status: string;
  additions: number;
  deletions: number;
  /// Unified-diff hunk text. `null` for binary files or when
  /// GitHub truncated past ~3MB.
  patch: string | null;
  /// Original filename when status === "renamed".
  previousFilename: string | null;
}

export interface CheckRun {
  name: string;
  /// `queued | in_progress | completed`.
  status: string;
  /// `success | failure | neutral | cancelled | skipped | timed_out |
  /// action_required`. `null` until status === "completed".
  conclusion: string | null;
  detailsUrl: string | null;
  startedAt: string | null;
  completedAt: string | null;
}

export type PrActionVerb =
  | "close"
  | "reopen"
  | "convertToDraft"
  | "markReadyForReview";

/// Strategy GitHub applies when merging the PR. Matches the backend
/// `MergeMethod` enum verbatim (serialised lowercase).
export type MergeMethod = "merge" | "squash" | "rebase";

export function integrationGetPrDetail(
  integrationType: string,
  owner: string,
  repo: string,
  number: number,
): Promise<PullRequestDetail> {
  return invoke<PullRequestDetail>("integration_get_pr_detail", {
    integrationType,
    owner,
    repo,
    number,
  });
}

export function integrationListPrCommits(
  integrationType: string,
  owner: string,
  repo: string,
  number: number,
): Promise<PrCommit[]> {
  return invoke<PrCommit[]>("integration_list_pr_commits", {
    integrationType,
    owner,
    repo,
    number,
  });
}

export function integrationListPrFiles(
  integrationType: string,
  owner: string,
  repo: string,
  number: number,
): Promise<PrFile[]> {
  return invoke<PrFile[]>("integration_list_pr_files", {
    integrationType,
    owner,
    repo,
    number,
  });
}

/// Fetch the Checks tab. GitHub uses `headSha` to look up
/// check-runs on the head commit; GitLab uses the MR `number` (iid)
/// to list pipelines. Pass both — providers ignore the irrelevant
/// argument, but each one is required by its own dispatch.
export function integrationListPrChecks(
  integrationType: string,
  owner: string,
  repo: string,
  headSha: string,
  number: number,
): Promise<CheckRun[]> {
  return invoke<CheckRun[]>("integration_list_pr_checks", {
    integrationType,
    owner,
    repo,
    number,
    headSha,
  });
}

export function integrationPrAction(
  integrationType: string,
  owner: string,
  repo: string,
  number: number,
  action: PrActionVerb,
): Promise<PullRequestDetail> {
  return invoke<PullRequestDetail>("integration_pr_action", {
    integrationType,
    owner,
    repo,
    number,
    action,
  });
}

/// Merge a PR via the matching provider's API. On success, optionally
/// drops the source branch (GitHub: separate `DELETE /git/refs`;
/// GitLab: `should_remove_source_branch` in the merge body). Returns
/// the post-mutation detail so the panel refreshes in one round-trip.
///
/// `squash` is GitLab-only: the GitLab API exposes squash as a flag
/// independent of merge method, so the form can offer "merge with merge
/// commit but squash all commits" combos. GitHub + Gitea derive squash
/// from `method === "squash"`.
export interface MergeOptions {
  method: MergeMethod;
  commitTitle?: string;
  commitMessage?: string;
  deleteSourceBranch: boolean;
  /// GitLab-only squash override. `undefined` = backend defaults
  /// (squash iff `method === "squash"`).
  squash?: boolean;
}

export function integrationMergePr(
  integrationType: string,
  owner: string,
  repo: string,
  number: number,
  options: MergeOptions,
): Promise<PullRequestDetail> {
  return invoke<PullRequestDetail>("integration_merge_pr", {
    integrationType,
    owner,
    repo,
    number,
    method: options.method,
    commitTitle: options.commitTitle ?? null,
    commitMessage: options.commitMessage ?? null,
    deleteSourceBranch: options.deleteSourceBranch,
    squash: options.squash ?? null,
  });
}

/// Rebase a GitLab MR's source branch onto target via
/// `PUT /merge_requests/:iid/rebase`. `skipCi` suppresses the pipeline
/// trigger that would otherwise fire on the rebased commits.
/// GitLab-only — calling against GitHub / Gitea returns an error from
/// the backend.
export function integrationRebaseMr(
  integrationType: string,
  owner: string,
  repo: string,
  number: number,
  skipCi: boolean,
): Promise<PullRequestDetail> {
  return invoke<PullRequestDetail>("integration_rebase_mr", {
    integrationType,
    owner,
    repo,
    number,
    skipCi,
  });
}

File: crates/yryvu-bridge/src/commands/pr_detail.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Tauri commands for the PR detail panel (#91 GitHub, #93 GitLab):
//! full-PR fetch, commits / files / checks tabs, and the action-button
//! cluster (close / reopen / convert-draft / mark-ready / merge).
//!
//! Each command routes by `ProviderFamily`. Supported providers
//! return the cross-provider shape; the rest bubble a typed error
//! the frontend surfaces as a toast.

use std::path::{Path, PathBuf};

use tauri::{AppHandle, Manager};

use crate::integrations::{
    self, CheckRun, MergeMethod, MergeRequest, MrAction, PrAction, PrCommit, PrFile,
    PullRequestDetail,
};

use super::integration_routing::{is_self_hosted, ProviderFamily};

fn sidecar_path(app: &AppHandle) -> Result<PathBuf, String> {
    let dir = app.path().app_local_data_dir().map_err(|e| e.to_string())?;
    Ok(dir.join("integrations.json"))
}

async fn load_auth(
    app: &AppHandle,
    integration_type: String,
) -> Result<integrations::AuthData, String> {
    let path = sidecar_path(app)?;
    tauri::async_runtime::spawn_blocking({
        let integration_type = integration_type.clone();
        move || integrations::get_integration(&path as &Path, &integration_type)
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())?
    .ok_or_else(|| format!("integration '{integration_type}' is not connected"))
}

fn pick_hostname(integration_type: &str, auth: &integrations::AuthData) -> Option<String> {
    is_self_hosted(integration_type)
        .then_some(auth.hostname.as_ref())
        .flatten()
        .cloned()
}

/// Verb the read-side panel emits. Same enum is used for both
/// providers; the per-family dispatch picks the concrete variant.
fn parse_action(action: &str) -> Result<(PrAction, MrAction), String> {
    match action {
        "close" => Ok((PrAction::Close, MrAction::Close)),
        "reopen" => Ok((PrAction::Reopen, MrAction::Reopen)),
        "convertToDraft" => Ok((PrAction::ConvertToDraft, MrAction::ConvertToDraft)),
        "markReadyForReview" => Ok((PrAction::MarkReadyForReview, MrAction::MarkReadyForReview)),
        other => Err(format!("unknown pr action: '{other}'")),
    }
}

fn unsupported_for_detail(integration_type: &str) -> String {
    format!("PR detail view is not implemented for provider '{integration_type}'")
}

#[tauri::command]
pub async fn integration_get_pr_detail(
    app: AppHandle,
    integration_type: String,
    owner: String,
    repo: String,
    number: u64,
) -> Result<PullRequestDetail, String> {
    let auth = load_auth(&app, integration_type.clone()).await?;
    let hostname = pick_hostname(&integration_type, &auth);
    let host = hostname.as_deref();
    match ProviderFamily::from_integration_type(&integration_type) {
        ProviderFamily::Github => {
            integrations::get_github_pr_detail(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitlab => {
            integrations::get_gitlab_mr_detail(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitea => {
            integrations::get_gitea_pr_detail(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        _ => Err(unsupported_for_detail(&integration_type)),
    }
}

#[tauri::command]
pub async fn integration_list_pr_commits(
    app: AppHandle,
    integration_type: String,
    owner: String,
    repo: String,
    number: u64,
) -> Result<Vec<PrCommit>, String> {
    let auth = load_auth(&app, integration_type.clone()).await?;
    let hostname = pick_hostname(&integration_type, &auth);
    let host = hostname.as_deref();
    match ProviderFamily::from_integration_type(&integration_type) {
        ProviderFamily::Github => {
            integrations::list_github_pr_commits(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitlab => {
            integrations::list_gitlab_mr_commits(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitea => {
            integrations::list_gitea_pr_commits(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        _ => Err(unsupported_for_detail(&integration_type)),
    }
}

#[tauri::command]
pub async fn integration_list_pr_files(
    app: AppHandle,
    integration_type: String,
    owner: String,
    repo: String,
    number: u64,
) -> Result<Vec<PrFile>, String> {
    let auth = load_auth(&app, integration_type.clone()).await?;
    let hostname = pick_hostname(&integration_type, &auth);
    let host = hostname.as_deref();
    match ProviderFamily::from_integration_type(&integration_type) {
        ProviderFamily::Github => {
            integrations::list_github_pr_files(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitlab => {
            integrations::list_gitlab_mr_files(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitea => {
            integrations::list_gitea_pr_files(&auth.token, host, &owner, &repo, number)
                .await
                .map_err(|e| e.to_string())
        }
        _ => Err(unsupported_for_detail(&integration_type)),
    }
}

/// Checks tab. GitHub uses the head SHA (the frontend already has
/// it); GitLab uses the MR iid for pipelines listing. The frontend
/// passes the SHA as the last argument either way — for GitLab the
/// SHA is unused. The parameter name stays `head_sha` to keep the
/// IPC surface stable across providers.
#[tauri::command]
pub async fn integration_list_pr_checks(
    app: AppHandle,
    integration_type: String,
    owner: String,
    repo: String,
    head_sha: String,
    number: Option<u64>,
) -> Result<Vec<CheckRun>, String> {
    let auth = load_auth(&app, integration_type.clone()).await?;
    let hostname = pick_hostname(&integration_type, &auth);
    let host = hostname.as_deref();
    match ProviderFamily::from_integration_type(&integration_type) {
        ProviderFamily::Github => {
            integrations::list_github_pr_checks(&auth.token, host, &owner, &repo, &head_sha)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitlab => {
            let iid = number.ok_or_else(|| {
                "GitLab checks require the MR iid; pass `number` from the panel".to_string()
            })?;
            integrations::list_gitlab_mr_pipelines(&auth.token, host, &owner, &repo, iid)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitea => {
            integrations::list_gitea_pr_checks(&auth.token, host, &owner, &repo, &head_sha)
                .await
                .map_err(|e| e.to_string())
        }
        _ => Err(unsupported_for_detail(&integration_type)),
    }
}

/// `action` is `close | reopen | convertToDraft | markReadyForReview`
/// (camelCase to match the frontend enum). Returns the post-mutation
/// PR detail so the panel can refresh in one round-trip.
#[tauri::command]
pub async fn integration_pr_action(
    app: AppHandle,
    integration_type: String,
    owner: String,
    repo: String,
    number: u64,
    action: String,
) -> Result<PullRequestDetail, String> {
    let (gh_action, gl_action) = parse_action(&action)?;
    let auth = load_auth(&app, integration_type.clone()).await?;
    let hostname = pick_hostname(&integration_type, &auth);
    let host = hostname.as_deref();
    match ProviderFamily::from_integration_type(&integration_type) {
        ProviderFamily::Github => {
            integrations::github_pr_action(&auth.token, host, &owner, &repo, number, gh_action)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitlab => {
            integrations::gitlab_mr_action(&auth.token, host, &owner, &repo, number, gl_action)
                .await
                .map_err(|e| e.to_string())
        }
        ProviderFamily::Gitea => {
            integrations::gitea_pr_action(&auth.token, host, &owner, &repo, number, gh_action)
                .await
                .map_err(|e| e.to_string())
        }
        _ => Err(unsupported_for_detail(&integration_type)),
    }
}

/// Merge a PR via the matching provider's API. `method` is one of
/// `merge | squash | rebase` (matches the frontend enum verbatim);
/// providers project the canonical enum onto their native semantics.
///
/// `squash` is an independent toggle honoured only by GitLab — the
/// other providers derive squash from `method == squash`. GitHub
/// drops the source branch via a follow-up `DELETE /git/refs/heads`
/// call; GitLab folds it into the merge body, Gitea defers it.
///
/// Argument count is gated by the Tauri IPC surface (each param maps
/// to one camelCase key on the JS side); refactoring into a wrapper
/// struct would force the frontend to serialise an envelope, costing
/// more than it saves.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn integration_merge_pr(
    app: AppHandle,
    integration_type: String,
    owner: String,
    repo: String,
    number: u64,
    method: String,
    commit_title: Option<String>,
    commit_message: Option<String>,
    delete_source_branch: bool,
    squash: Option<bool>,
) -> Result<PullRequestDetail, String> {
    let auth = load_auth(&app, integration_type.clone()).await?;
    let hostname = pick_hostname(&integration_type, &auth);
    let host = hostname.as_deref();
    let parsed_method = match method.as_str() {
        "merge" => MergeMethod::Merge,
        "squash" => MergeMethod::Squash,
        "rebase" => MergeMethod::Rebase,
        other => return Err(format!("unknown merge method: '{other}'")),
    };
    let request = MergeRequest {
        method: parsed_method,
        commit_title,
        commit_message,
        squash,
    };
    let family = ProviderFamily::from_integration_type(&integration_type);
    let detail = match family {
        ProviderFamily::Github => {
            integrations::github_merge_pr(&auth.token, host, &owner, &repo, number, request)
                .await
                .map_err(|e| e.to_string())?
        }
        ProviderFamily::Gitea => {
            integrations::gitea_merge_pr(&auth.token, host, &owner, &repo, number, request)
                .await
                .map_err(|e| e.to_string())?
        }
        ProviderFamily::Gitlab => integrations::gitlab_merge_mr(
            &auth.token,
            host,
            &owner,
            &repo,
            number,
            request,
            delete_source_branch,
        )
        .await
        .map_err(|e| e.to_string())?,
        _ => {
            return Err(unsupported_for_detail(&integration_type));
        }
    };
    if delete_source_branch && matches!(family, ProviderFamily::Github) {
        // GitHub only: branch delete is a separate `DELETE /git/refs`
        // call. Failure here doesn't fail the merge (the user's
        // primary intent already succeeded — branch lingers, audit
        // log carries the reason).
        if let Err(err) =
            integrations::github_delete_branch(&auth.token, host, &owner, &repo, &detail.head_ref)
                .await
        {
            eprintln!("github delete branch failed (branch survives): {err}");
        }
    }
    Ok(detail)
}

/// Rebase a GitLab MR's source branch onto target via
/// `PUT /merge_requests/:iid/rebase`. `skip_ci` suppresses the
/// pipeline trigger that would otherwise fire on the rebased commits.
/// GitLab returns 202 immediately; we re-fetch the detail so the
/// panel sees the freshest state.
#[tauri::command]
pub async fn integration_rebase_mr(
    app: AppHandle,
    integration_type: String,
    owner: String,
    repo: String,
    number: u64,
    skip_ci: bool,
) -> Result<PullRequestDetail, String> {
    let auth = load_auth(&app, integration_type.clone()).await?;
    let hostname = pick_hostname(&integration_type, &auth);
    let host = hostname.as_deref();
    match ProviderFamily::from_integration_type(&integration_type) {
        ProviderFamily::Gitlab => {
            integrations::gitlab_rebase_mr(&auth.token, host, &owner, &repo, number, skip_ci)
                .await
                .map_err(|e| e.to_string())
        }
        _ => Err(format!(
            "rebase action is GitLab-only (got '{integration_type}')"
        )),
    }
}
